compiler/reader.go

```diff
@@ -47,60 +50,94 @@ func initializeInfoCache() {
 	}
 }
 
+// EnableFileCache turns on file caching.
 func EnableFileCache() {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	fileCacheEnable = true
 }
 
+// EnableInfoCache turns on parsed info caching.
 func EnableInfoCache() {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	infoCacheEnable = true
 }
 
+// DisableFileCache turns off file caching.
 func DisableFileCache() {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	fileCacheEnable = false
 }
 
+// DisableInfoCache turns off parsed info caching.
 func DisableInfoCache() {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	infoCacheEnable = false
 }
 
+// RemoveFromFileCache removes an entry from the file cache.
 func RemoveFromFileCache(fileurl string) {
+	fileCacheMutex.Lock()
+	defer fileCacheMutex.Unlock()
 	if !fileCacheEnable {
 		return
 	}
 	initializeFileCache()
 	delete(fileCache, fileurl)
 }
 
+// RemoveFromInfoCache removes an entry from the info cache.
 func RemoveFromInfoCache(filename string) {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
 	if !infoCacheEnable {
 		return
 	}
 	initializeInfoCache()
 	delete(infoCache, filename)
 }
 
+// GetInfoCache returns the info cache map.
 func GetInfoCache() map[string]interface{} {
+	infoCacheMutex.Lock()
+	defer infoCacheMutex.Unlock()
```

Comment:
Is the locking only necessary when `initializeInfoCache()` is called? Or is it also necessary to return the reference?

If the former, please move the locking/unlocking inside the conditional.

Reply:
I think we should protect the test of infoCache, but generally should clarify that the locking strategy here is to make all of the public functions in this directory exclusive. So every capitalized function begins with a lock/defer unlock and these functions never call other capitalized functions - accesses inside the uncapitalized (private) functions are unsynchronized because they are synchronized at the higher level of the public function interfaces.

Reply:
I'll add a comment to clarify this in the code.